src/mmds/src/token.rs

```diff
@@ -0,0 +1,367 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use std::{fmt, io};
+
+/// Path to token.
+pub const PATH_TO_TOKEN: &str = "/api/token";
+/// Length of the token.
+pub const TOKEN_LEN: usize = 56;
+/// Minimum lifetime of token.
+pub const MIN_TOKEN_TTL_SECONDS: u32 = 1;
+/// Maximum lifetime of token.
+pub const MAX_TOKEN_TTL_SECONDS: u32 = 21600;
+/// Randomness pool file path.
+const RANDOMNESS_POOL: &str = "/dev/urandom";
+/// Maximum number of sessions.
+const MAX_SESSIONS: usize = 100;
+
+#[derive(Debug)]
+pub enum TokenError {
+    /// Failed to open `/dev/urandom` file.
+    EntropyPool(io::Error),
+    /// Time to live value for token is invalid.
+    InvalidTtlValue(u32),
+    /// Failed to read from `/dev/urandom` file.
+    ReadFromEntropyPool(io::Error),
+}
+
+impl fmt::Display for TokenError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &*self {
+            TokenError::EntropyPool(err) => write!(f, "Failed to open entropy pool: {}.", err),
+            TokenError::InvalidTtlValue(value) => write!(
+                f,
+                "Invalid time to live value provided for token: {}. \
+                Please provide a value between {} and {}.",
+                value, MIN_TOKEN_TTL_SECONDS, MAX_TOKEN_TTL_SECONDS,
+            ),
+            TokenError::ReadFromEntropyPool(err) => {
+                write!(f, "Failed to read from entropy pool: {}.", err)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum TokenStoreError {
+    /// Failed to generate session token.
+    TokenGenerationError(TokenError),
+    /// Session tokens limit exceeded.
+    TokensLimitExceeded,
+}
+
+impl fmt::Display for TokenStoreError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &*self {
+            TokenStoreError::TokenGenerationError(err) => {
+                write!(f, "Failed to generate session token: {}", err)
+            }
+            TokenStoreError::TokensLimitExceeded => {
+                write!(f, "Tokens limit exceeded: {}.", MAX_SESSIONS)
+            }
+        }
+    }
+}
+
+/// Structure for information regarding token lifetime.
+#[derive(Clone)]
+pub struct TokenInfo {
+    ttl_seconds: u32,
+    start_time: DateTime<Utc>,
+}
+
+impl TokenInfo {
+    /// Create a new session Token for MMDS requests.
+    fn new(ttl_seconds: u32) -> Self {
+        TokenInfo {
+            ttl_seconds,
+            start_time: Utc::now(),
+        }
+    }
+
+    /// Return the expiry date of the token.
+    fn expiry(&self) -> DateTime<Utc> {
+        self.start_time
+            .checked_add_signed(Duration::seconds(self.ttl_seconds as i64))
+            .expect("Failed to compute token expiry date.")
+    }
+
+    /// Specifies whether the token has expired.
+    pub fn has_expired(&self) -> bool {
+        Utc::now()
+            .signed_duration_since(self.expiry())
+            .num_milliseconds()
+            >= 0
+    }
```

Comment:
We can avoid `chrono` by using `std::time::{Duration, Instant}`. Smth like:

```suggestion
    fn new(ttl_seconds: u32) -> Self {
        TokenInfo {
            ttl: Duration::from_secs(ttl_seconds),
            start_time: Instant::now(),
        }
    }

    /// Specifies whether the token has expired.
    pub fn has_expired(&self) -> bool {
        self.start_time.elapsed().as_millis() >= self.ttl
    }
```

Reply:
You are right, we don't need `chrono`.